docs: improve GraphQL with example and better format

TomChv opened the pull request:
Improve GraphQL doc with more examples.
Generator templates correctly format examples.

Resolves: #4323
Signed-off-by: Tom Chauveau <tom.chauveau@epitech.eu>

slumbering commented on sdk/nodejs/api/client.gen.ts:
```diff
@@ -495,9 +586,10 @@ The command being executed WILL BE GRANTED FULL ACCESS TO YOUR HOST FILESYSTEM.
    * @param path Host's destination path.
-Path can be relative to the engine's workdir or absolute.
+   * Path can be relative to the engine's workdir or absolute.
+   * @example ./tarball
```
However, for class methods it's displayed correctly 😃 
Could we have a more explicit example though ?
```ts
/**
* @example
* export("./tarball")
*/
```

Before:
![image](https://user-images.githubusercontent.com/1186424/212037270-fbce8cc5-cc9c-4333-aef2-c74224ffa278.png)

After:
![image](https://user-images.githubusercontent.com/1186424/212037311-8d928a2b-2ed7-46a8-b713-726261e6a467.png)

TomChv replied:
Function call example instead of parameter example, I understand!

TomChv replied:
Fixed in https://github.com/dagger/dagger/pull/4367/commits/2a0ebdce4d802f8666d5963ca8963641169fd915

slumbering replied:
Much better thanks 😄 ! However it display many example blocks now instead of one grouping snippets. Let move forward though and addressed that in another issue. Thanks

![image](https://user-images.githubusercontent.com/1186424/212274263-f21ddd1a-aba5-4933-a5a9-481d434b8841.png)

slumbering reviewed (commented):
Thanks @TomChv, awesome improvements. Docs are getting better day after day :)
Some comments about how examples are displayed.

slumbering commented on sdk/nodejs/api/client.gen.ts:
```diff
@@ -219,26 +271,41 @@ export type DirectoryPipelineOpts = {
 export type DirectoryWithFileOpts = {
+  /**
+   * Permission given to the copied file.
+   * Default to 0644.
```
Is it possible to use `@default` instead of `Default`. If not it's fine.

TomChv replied:
Possible but I would like to keep it for a follow-up PR if possible!

TomChv replied:
Fixed in https://github.com/dagger/dagger/pull/4367/commits/2a0ebdce4d802f8666d5963ca8963641169fd915

slumbering replied:
I don't see any updates, Default is still the same. But honestly like I've said it's not a big deal and like you've said it can be addressed in a following PR.

slumbering commented on sdk/nodejs/api/client.gen.ts:
```diff
@@ -1353,6 +1515,12 @@ Defaults to './Dockerfile'.
+   * @param path Location of the written file.
+   * @example /file.txt
+   * @param contents Content of the written file.
+   * @example Hello world
```
When you have more than one param, it create an example per param...

![image](https://user-images.githubusercontent.com/1186424/212038549-68b067bf-aba8-41dc-8b4c-8099af0d5aab.png)

TomChv replied:
Ok I think I must change the format for the function to have only one example that groups the function call instead of an example by parameters.
I'll work on it.

TomChv replied:
Fixed in https://github.com/dagger/dagger/pull/4367/commits/2a0ebdce4d802f8666d5963ca8963641169fd915

slumbering replied:
same comment as here : https://github.com/dagger/dagger/pull/4367#discussion_r1069081894

TomChv commented on sdk/nodejs/api/client.gen.ts:
```diff
@@ -53,7 +53,16 @@ class BaseClient {
 export type BuildArg = {
+  /**
+   * The build argument name.
+   * @example MODE
```
I'll try to find a way to format it good!

TomChv replied:
The problem with rolling back only for this case is that it creates inconsistency. I can easily handle that in the codegen but that will be inconsistent.
Instead, we shall maybe try to fix the theme to handle this case correctly. I'll see if it's possible.

TomChv replied:
Ok I found a way! instead of giving an example in the field, we can give it in the exec call so it's appropriately displayed, and then we can remove the example for the field.
So only methods function calls will be documented with examples that make sense in a doc.

TomChv replied:
Fixed in https://github.com/dagger/dagger/pull/4367/commits/2a0ebdce4d802f8666d5963ca8963641169fd915

slumbering commented:
LGTM for nodejs. @helderco @dolanor, since it's also impacting go and python side could you give a quick review please ?

dolanor commented on codegen/generator/nodejs/templates/src/method_comment.ts.tmpl:
```diff
@@ -4,26 +4,56 @@
 		{{- range $optionals }}
 			{{- if .Description }}
-   * @param opts.{{ .Name }} {{ .Description }} 
+		        {{- /* Reference current arg to access it in range */ -}}
```
indent 1 level under the `if` block

TomChv replied:
Good catch! I'm waiting for feedback on https://github.com/dagger/dagger/pull/4367#issuecomment-1382619290 before changing this.

dolanor replied:
I think this is not impacted by the linked comment and should be indented 1 level under the `if`

helderco commented on core/schema/cache.graphqls:
```diff
@@ -2,11 +2,13 @@
+  """
+  Constructs a cache volume for a given cache key.
+  Example:
+  cacheVolume("myapp-cache")
```
We should avoid referencing field and parameter names in the documentation because other SDKs have different naming conventions. This won't look right in Go or Python (`CacheVolume` and `cache_volume` respectively).

TomChv replied:
See https://github.com/dagger/dagger/pull/4367#issuecomment-1382619290

helderco commented on core/schema/cache.graphqls:
```diff
@@ -2,11 +2,13 @@
+  """
   cacheVolume(
-    """
-    A string identifier to target this cache volume (e.g. "myapp-cache").
```
The `e.g.` was fine here.

TomChv replied:
I put examples on the function call to be more explicit than the field parameter.
See my comments [here](https://github.com/dagger/dagger/pull/4367#issuecomment-1382619290) for details and lmk what you would prefer.

dolanor commented on sdk/nodejs/api/client.gen.ts:
```diff
@@ -219,26 +254,36 @@ export type DirectoryPipelineOpts = {
 
   /**
    * Include only artifacts that match the given pattern.
-   * (e.g. ["app/", "package.*"]).
```
why did this disappear?

TomChv replied:
See https://github.com/dagger/dagger/pull/4367#issuecomment-1382619290

helderco commented on core/schema/container.graphqls:
```diff
@@ -23,27 +24,39 @@ type Container {
+  Example:
+  build("directory-id")
+  build("directory-id", { dockerfile: "./frontend.Dockerfile" })
+  build("directory-id", { buildArgs: [{ name: "MODE", value: "production }] })
```
These examples are specific to the NodeJS SDK, they don't work even for GraphQL.

Besides, the `"directory-id"` value is misleading as a proper value for a `DirectoryID`.

TomChv replied:
It's a function call example yeah, it's abstract.

helderco replied:
You can also use a variable (e.g., `$directoryID`) or a truncated —but more correct— value (e.g., `"eyJmc...9fQ=="`).

dolanor commented on sdk/nodejs/api/client.gen.ts:
```diff
@@ -455,9 +523,13 @@ Defaults to './Dockerfile'.
    * @param opts.redirectStdout Redirect the command's standard output to a file in the container.
    * @param opts.redirectStderr Redirect the command's standard error to a file in the container.
    * @param opts.experimentalPrivilegedNesting Provide dagger access to the executed command.
-Do not use this option unless you trust the command being executed.
```
Good catch on this! :+1:

TomChv replied:
Thanks :rocket:

dolanor reviewed (changes requested):
Here is the first pass. I haven't finished yet.
I want to try on my machine and check some edge cases.

helderco reviewed (dismissed):
I appreciate having more documentation 👍 but I'm less keen in the way the examples are constructed because of the different naming conventions.

We need a doc style guide for the API with conventions and rules to follow. Would be interesting to find some common ground between all SDKs and GraphQL itself. Probably hard but here’s some ideas:
- Short title first, then blank line, then description.
- End sentences in a period, in proper english, even if it's just a short title.
- Avoid referencing field and parameter names due to differences in naming conventions.
- There’s been an exception to the previous one which relates to deprecations. Referencing the new field in backticks have allowed the SDKs to rename its contents into that language’s naming convention.

TomChv commented:
> We need a doc style guide for the API with conventions and rules to follow. Would be interesting to find some common ground between all SDKs and GraphQL itself. Probably hard but here’s some ideas:
>
> Short title first, then blank line, then description.
> End sentences in a period, in proper english, even if it's just a short title.
> Avoid referencing field and parameter names due to differences in naming conventions.
> There’s been an exception to the previous one which relates to deprecations. Referencing the new field in backticks have allowed the SDKs to rename its contents into that language’s naming convention.

Before making any more changes, I think you're right, we should define a standard.

> Short title first, then blank line, then description.

Totally agree, I'm thinking about a short line desc and if necessary -> blank line + big description.

> End sentences in a period, in proper english, even if it's just a short title.

Not sure I get that one, can you give me an example, please?

> Avoid referencing field and parameter names due to differences in naming conventions.

I agree that it's not optimal but it produce a really good result for NodeJS documentation.
Instead of avoiding referencing, we could update generators to correctly format the name of the function?

In GQL, we keep the same name as the method and it's up to the generator to correctly format it, like we do for method name? WDYT?
Examples will not be perfect, for instant in Go it will lack of `[]string` for an array but it's still better than nothing.

Another solution is to remove function call example and only stick to `e.g` format but that will not be possible to have `@example` tag in Typescript, since examples must be at the function level, not on parameter/opts.

I'm fine with both solutions, the first one requires more work but will give a better doc with explicit examples. The other one is simple but we will not have an explicit one.

TomChv commented on codegen/generator/nodejs/templates/src/method_comment.ts.tmpl:
```diff
@@ -4,26 +4,56 @@
 	{{- range $required }}
 		{{- if .Description }}
-   * @param {{ .Name }} {{ .Description }}
+		{{- /* Reference current arg to access it in range */ -}}
```
Good catch! I'm waiting for feedback on https://github.com/dagger/dagger/pull/4367#issuecomment-1382619290 before changing this.

dolanor replied:
I think this is not impacted by the linked comment and should be indented 1 level under the `if`

slumbering commented:
> > Short title first, then blank line, then description.
> 
> Totally agree, I'm thinking about a short line desc and if necessary -> blank line + big description.

Also agree with that 👍 

> 
> > End sentences in a period, in proper english, even if it's just a short title.
> 
> Not sure I get that one, can you give me an example, please?

I guess what it means is we should add a period for titles, short descriptions, long descriptions etc. No particular example, it's just to have a exhaustive rules list. It's my understanding.

> 
> > Avoid referencing field and parameter names due to differences in naming conventions.
> 
> I agree that it's not optimal but it produce a really good result for NodeJS documentation. Instead of avoiding referencing, we could update generators to correctly format the name of the function?
> 
> In GQL, we keep the same name as the method and it's up to the generator to correctly format it, like we do for method name? WDYT? Examples will not be perfect, for instant in Go it will lack of `[]string` for an array but it's still better than nothing.
> 
> Another solution is to remove function call example and only stick to `e.g` format but that will not be possible to have `@example` tag in Typescript, since examples must be at the function level, not on parameter/opts.
> 
> I'm fine with both solutions, the first one requires more work but will give a better doc with explicit examples. The other one is simple but we will not have an explicit one.

Like @TomChv said, fields names are the same as the one in GraphQL. There're not relative to any SDKs.
However, parameters are clearly for the Node.js SDK...and  It feels weird to use those in a `.graphql` file.
I think it would make more sense to declare those parameters in a GraphQL way and leave each SDK the responsibility to format them correctly.

Before
```
dockerBuild()
dockerBuild({ dockerfile: "frontend.Dockerfile" })
dockerBuild({ platform: "linux/amd64" })
dockerBuild({ platform: "linux/amd64", buildArgs: [{ name: "MODE", value: "production" }] })
```

After
```
dockerBuild(
  dockerfile: "frontend.Dockerfile"
  platform: "linux/amd64"
  buildArgs: [{ name: "MODE", value: "production" }], 
  target:"/"
)
```

dolanor commented:
> > End sentences in a period, in proper english, even if it's just a short title.
>
> Not sure I get that one, can you give me an example, please?
I think, @helderco just means, end your sentences with a `.`, like in plain english.

My point of view about this:
- for us, it's less confusing to read doc that is not totally idiomatic for the SDK/tech because we know it might be some incompatibilities. For newcomers to the specific SDK they know, it will look alien and might lose them.
- if we want to have a real full compatibility, we will have to have some real doc parser that can extract function parameter names, and can reformat them. That might be a lot of work and error-prone.
  + I think the approach that @helderco talked about with surrounding param names with backticks ` `` ` could work in the description body.
  + For examples, this would be more tricky, but I guess we could parse the example paragraph from the description as graphQL, and translate that as the SDK code we generate for. I think this one would be the most painful and error prone.

> Avoid referencing field and parameter names due to differences in naming conventions.

That wouldn't work for Go, as it IS the way to create documentation for a function. It is part of the documentation description. I believe surrounding the params with ` `` ` would work (even though not idiomatic)

I found the format for giving example in Go, though. It is just a comment line prefixed by a ` `
=> https://tip.golang.org/doc/comment#code

helderco commented:
> > Short title first, then blank line, then description.
> 
> Totally agree, I'm thinking about a short line desc and if necessary -> blank line + big description.

That's what I mean. Ideally the first line should be as short as the max lengths allow in styling rules so as not to break it into two. In the API currently there's never a break between the first line and the rest of the description.

One liners are already ok:

```graphql
  "A unique identifier for this container."
  id: ContainerID!
```

But even if there's two lines, it should be broken:

```diff
    """
    Identifiers for other platform specific containers.
+
    Used for multi-platform image.
    """
    platformVariants: [ContainerID!]
```

Or when you have "Defaults to", "Formatted as", "Example", etc...

```diff
    """
    Path to the Dockerfile to use.
+ 
    Defaults to './Dockerfile'.
    """
    dockerfile: String
```

---

> > End sentences in a period, in proper english, even if it's just a short title.
> 
> Not sure I get that one, can you give me an example, please?

You've already fixed this in:
- https://github.com/dagger/dagger/pull/4240

Example:

```diff
- "A global cache volume identifier"
+ "A global cache volume identifier."
 scalar CacheID
```

---

> > Avoid referencing field and parameter names due to differences in naming conventions.
> 
> I agree that it's not optimal but it produce a really good result for NodeJS documentation. Instead of avoiding referencing, we could update generators to correctly format the name of the function?

If that's all it was (i.e., name of the function) it would be fine, but if you take into account the different ways **parameters** are called, I don't think this is feasible.

---

> In GQL, we keep the same name as the method and it's up to the generator to correctly format it, like we do for method name? WDYT? Examples will not be perfect, for instant in Go it will lack of `[]string` for an array but it's still better than nothing.

I disagree. It's better than nothing for NodeJS, but incorrect in other languages isn't better than nothing to me. Besides, if there should be one language to optimize for it should be GraphQL, not NodeJS.

---

> Another solution is to remove function call example and only stick to `e.g` format but that will not be possible to have `@example` tag in Typescript, since examples must be at the function level, not on parameter/opts.

We can't have everything 🙂

helderco commented:
> * if we want to have a real full compatibility, we will have to have some real doc parser that can extract function parameter names, and can reformat them. That might be a lot of work and error-prone.

@dolanor To give more clarity to my previous point, this is exactly the thing that would be needed but I **don't think it's worth it**. 

Even if for NodeJS it becomes something worth doing for, it would force others to do it as well.

---

> I think the approach that @helderco talked about with surrounding param names with backticks `` could work in the description body.

This is only used now to reference other fields. If you're documenting the current field, it's easy enough to replace the current field name in the description as well, without the backticks. 

The problem again is the different ways the parameters are called.

TomChv commented:
Thanks everyone for your feedback! I updated the PR's name to fit the changes on this PR.

What I extracted from this is:

**Format description**
- One short line that briefly describes the function or type and ends with a dot (".").
  - For function: must start with a 3rd person verb.
- For any additional doc contents (or example, default value etc...): leave an empty line after the short line.

**Field example**
- To give enough doc on any SDK but still not be specific, it should follow the syntax (`e.g., <example>)`
- It should be done before `Default` and `Example`
- We may break a line if necessary or put it at the end of the sentence.

**Default value**
- They shall be introduced with `Default:` and written before `Example:` so it's included in generated docs).

**Function example**
- They shall be introduced with an `Example:` line so it's easily parses (and so ignores for now in SDK doc generation)
- They shall be formatted as GraphQL query

**Example compliant with the above spec**

```gql
// An OCI container... lorem ipsum.
//
// Lorem ipsum...
type Container {
    """
    Retrieves a container from a remote registry.

    Example:
    from({ address: "index.docker.io/alpine" })
    from({ address: "index.docker.io/golang:1.19", opts: [{ name: "no-cache", value: "true"  }] })
    """
    from(
       """
       The address of the image to pull.

       Formatted as {registry host}/{owner}/{image}/{tag}.
       (e.g., "index.docker.io/dagger/dagger")
       """
       address: String!,

      "Additional options to use on the pull."
      opts: [ContainerFromOpt!]
   ): Container!
}

"Optional arguments to pull a container."
input ContainerFromOpt {
   "The option's name (e.g., `my-option`)."
  name: String!

  "The option's value (e.g., `VALUE`). "
  value: String!
}
```

helderco commented:
## Backticks

> * For one line field example, we can't use `"` so we will use backticks **"`"**.

Backticks are currently only used to replace field name references to current language's naming conventions so I prefer to avoid adding more cases if we can. So far I haven't found the need.

For one liners you can still use the triple quotes. Instead of doing it only when needed, I suggest making it a convention. It also helps with git diffs in case the content changes (e.g., turning it multiline or needing double quotes in an example). 

> **Note**:
> Add it to style guide.

```graphql
    """
    The option's name (e.g., "my-option").
    """
    name: String!

    """
    Additional options to use on the pull.
    """
    opts: [ContainerFromOpt!]
```

## Function example

Let me see if I understand what you're suggesting correctly:
- Put examples under a specific `Example:` keyword;
- SDKs can find that string and delete everything from there to the end of the description;
- Must be put last because of the previous point.

Is that right? Have you tried to see how this looks in GraphiQL? Do you know of any other GraphQL API that documents examples like this in the field descriptions?

---

Nit: your last example isn't valid against our API since `from` doesn't have `opts`. I know you weren't focused on correctness but it's still interesting to get an example from the API that has optional arguments (e.g., `Container.build`) as part of a style guide.

```graphql
from(address: "index.docker.io/golang:1.19", opts: [{ name: "no-cache", value: "true"  }])
```

helderco commented:
Just realized I didn't give an opinion on my last "function example" comment.

As the intention with this was to improve the NodeJS SDK's reference docs, are you planning on parsing those GraphQL parameters and converting for the NodeJS SDK? If every SDK is to just remove that block, it'll only show in the playground.

My preference is to just give examples to possible input values (function parameters) if that helps clarify usage, but not full parametrized field calls (function calls).

I'll remove the "changes required" review if you cut that block from the SDKs but I'd get more opinions here (from engine and Go SDK folks). I also asked about other APIs as I'm genuinely curious how this is usually done (from the point of view of GraphQL alone). Haven't looked.

TomChv commented:
> Let me see if I understand what you're suggesting correctly:
>
> Put examples under a specific Example: keyword;
> SDKs can find that string and delete everything from there to the end of the description;
> Must be put last because of the previous point.
> Is that right? Have you tried to see how this looks in GraphiQL? Do you know of any other GraphQL API that documents examples like this in the field descriptions?

Exactly!



> @TomChv, as the intention with this was to improve the NodeJS SDK's reference docs, are you planning on parsing those GraphQL parameters and converting for the NodeJS SDK? If every SDK is to just remove that block, it'll only show in the playground.

Yeah the idea for now is to ignore those examples on SDK generation, I'll check what is the standard convention for GraphQL example.
I think it's still useful even if it's only on the playground, we may use those example later to format it on our SDKs and use it.

> My preference is to just give examples to possible input values (function parameters) if that helps clarify usage, but not full parametrized field calls (function calls).

For SDK that will be the case, but GQL doc will benefits of full examples :D

slumbering commented:
@dolanor @helderco could you have a last look on this one please ?

helderco reviewed (commented):
Some comments are relevant for multiple places but I only commented on one. For example:
- There's some summary lines that are too long;
- Missing blank lines between summary or other section, e.g., "Example";
- Most of the field-level examples are redundant with their parameters' docs;

I still don't see much value in going through the trouble of adding full examples to the top of fields. The idea was that since SDKs are striping the example, it would at least be useful for the playground and API reference users.

I took some screenshots to analyze this.

First, in the reference, the formatting isn't very good.

<img width="857" alt="Screenshot 2023-01-24 at 16 35 27" src="https://user-images.githubusercontent.com/174525/214379984-e41c092b-d6c3-427e-bb00-81cbd945c547.png">

It's not much better in the playground:

<img width="313" alt="Screenshot 2023-01-24 at 16 35 01" src="https://user-images.githubusercontent.com/174525/214379618-d5e9a1ea-e7f5-4b2e-a549-336b890fe2eb.png">

Granted, GraphQL descriptions can be written in markdown so we can turn that block into a code block:

<img width="318" alt="Screenshot 2023-01-24 at 16 41 46" src="https://user-images.githubusercontent.com/174525/214380408-963c0391-7e72-4853-85ef-4177c7f20b23.png">

There's still a lot of horizontal scrolling in a small width panel so it's not great.

The reference can also be improved to render the examples better. Although there's already some support for examples there that are complete and linked to the playground.

Here's an example:

<img width="599" alt="Screenshot 2023-01-24 at 16 45 00" src="https://user-images.githubusercontent.com/174525/214382203-2f1d518d-128e-4925-a74c-356691f53e56.png">

So I don't see much value in adding the full examples to the top. Maybe there's one case or another where it helps clarify in a big way how you use the parameters together but since every field has them, no such case has stood out to me.

In summary, the reference docs can have more complete examples in the form of queries, linked to the playground. The SDKs need to strip them out because of different signatures. And in the playground it doesn't look that great but you can also play with multiple values there and see the effect pretty fast.

TomChv commented:
> The reference can also be improved to render the examples better. Although there's already some support for examples there that are complete and linked to the playground.

You're right, I'll remove those examples at the top of the function now. I can work on them directly on the reference.

I'll take care of your change request so we can merge that PR after that.
I may work again on  GQL doc examples later, but I think for now, my changes shall be enough to improve the DX.

helderco commented on core/schema/container.graphqls:
```diff
@@ -61,121 +71,359 @@ type Container {
+  """
+  withWorkdir(
+    """
+    The path to set to the working directory (e.g., "/app").
```
```suggestion
    The path to set as the working directory (e.g., "/app").
```

helderco commented on core/schema/container.graphqls:
```diff
@@ -61,121 +71,359 @@ type Container {
+
+  """
+  Retrieves this container plus the given environment variable.
+
```
Extra new line here.

helderco commented on core/schema/container.graphqls:
```diff
@@ -61,121 +71,359 @@ type Container {
+  """
+  withEnvVariable(
+    """
+    The identifier of the environment variable (e.g., "HOST").
```
```suggestion
    The name of the environment variable (e.g., "HOST").
```

helderco commented on core/schema/container.graphqls:
```diff
@@ -61,121 +71,359 @@ type Container {
+  """
+  withSecretVariable(
+    """
+    The identifier of the secret variable (e.g., "API_SECRET").
```
```suggestion
    The name of the secret variable (e.g., "API_SECRET").
```

helderco commented on core/schema/container.graphqls:
```diff
@@ -61,121 +71,359 @@ type Container {
+  """
+  withoutEnvVariable(
+    """
+    The identifier of the environment variable (e.g., "HOST").
```
```suggestion
    The name of the environment variable (e.g., "HOST").
```

helderco commented on core/schema/container.graphqls:
```diff
@@ -61,121 +71,359 @@ type Container {
+    exclude: [String!],
+
+    """
+    Patterns to include in the written directory (e.g., ["*.go", "go.mod", "go.sum").
```
Typo:

```suggestion
    Patterns to include in the written directory (e.g., ["*.go", "go.mod", "go.sum"]).
```

helderco commented on core/schema/container.graphqls:
```diff
@@ -61,121 +71,359 @@ type Container {
+    """
+    path: String!,
+
+    "Identifier of the socket to forward."
```
```suggestion
    """
    Identifier of the socket to forward."
    """
```

TomChv replied:
Where does not? I can't see the error here

helderco replied:
I just think it's better to always use the triple quotes instead of when you have a quote inside the description like in the examples. So it becomes the convention to always use `"""` for these descriptions.

dolanor replied:
As it will improving diffing if we switch from 1-line to multi-lines description, I can agree with it too :)
Let's just use `"""` for those GraphQL description.

helderco replied:
> As it will improving diffing

Exactly, I mentioned it in https://github.com/dagger/dagger/pull/4367#issuecomment-1398350547.

helderco commented on core/schema/container.graphqls:
```diff
@@ -228,14 +486,18 @@ type Container {
+  Publishes this container as a new image to the specified address.
+
+  Publish returns a fully qualified ref.
+  It can also publishes platformVariants.
```
```suggestion
  It can also publish platform variants.
```

helderco commented on core/schema/container.graphqls:
```diff
@@ -246,14 +508,16 @@ type Container {
   Writes the container as an OCI tarball to the destination file path on the host for the specified platformVariants.
+
   Return true on success.
+  It can also publishes platformVariants.
```
```suggestion
  It can also publish platform variants.
```

helderco commented on core/schema/container.graphqls:
```diff
@@ -246,14 +508,16 @@ type Container {
 
   """
   Writes the container as an OCI tarball to the destination file path on the host for the specified platformVariants.
```
```suggestion
  Writes the container as an OCI tarball to the destination file path on the host for the specified platform variants.
```

helderco commented on core/schema/container.graphqls:
```diff
@@ -281,7 +545,11 @@ type Label {
   value: String!
 }
 
+"Key value object that represents a build argument."
 input BuildArg {
+  "The build argument name."
   name: String!
+
+  "The build argument value."
```
Nit: not using `"""` convention.

helderco commented on core/schema/directory.graphqls:
```diff
@@ -14,76 +14,197 @@ type Directory {
-  "Retrieves this directory with all file/dir timestamps set to the given time, in seconds from the Unix epoch."
-  withTimestamps(timestamp: Int!): Directory!
+  """
+  Retrieves this directory with all file/dir timestamps set to the given time, in seconds from the Unix epoch.
```
The part "in seconds from the Unix epoch" is redundant with the `timestamp` parameter doc.

```suggestion
  Retrieves this directory with all file/dir timestamps set to the given time.
```

helderco commented on core/schema/directory.graphqls:
```diff
@@ -14,76 +14,197 @@ type Directory {
+  """
+  withTimestamps(
+    """
+    Timestamp to set dir/files in (e.g., 1672531199).
+
+    Formatted in seconds following Unix epoch.
```
```suggestion
    Timestamp to set dir/files in.

    Formatted in seconds following Unix epoch (e.g., 1672531199).
```

helderco commented on core/schema/directory.graphqls:
```diff
@@ -14,76 +14,197 @@ type Directory {
+  ): Directory!
 }
 
+"Key value object that represents a build argument."
```
```suggestion
"""
Key value object that represents a build argument.
"""
```

helderco commented on core/schema/file.graphqls:
```diff
@@ -20,9 +20,25 @@ type File {
-  "Retrieves this file with its created/modified timestamps set to the given time, in seconds from the Unix epoch."
-  withTimestamps(timestamp: Int!): File!
+  """
+  Retrieves this file with its created/modified timestamps set to the given time, in seconds from the Unix epoch.
```
```suggestion
  Retrieves this file with its created/modified timestamps set to the given time.
```

helderco commented on core/schema/file.graphqls:
```diff
@@ -20,9 +20,25 @@ type File {
+  Retrieves this file with its created/modified timestamps set to the given time, in seconds from the Unix epoch.
+"""
+  withTimestamps(
+    """
+    Timestamp to set dir/files in (e.g., 1672531199).
+
+    Formatted in seconds following Unix epoch.
+    """
```
```suggestion
    """
    Timestamp to set dir/files in.

    Formatted in seconds following Unix epoch (e.g., 1672531199).
    """
```

helderco commented on core/schema/git.graphqls:
```diff
@@ -1,24 +1,57 @@
+    """
+    git(
+        """
+        Url of the git repository.
```
```suggestion
        Url of the git repository.

```

helderco commented on core/schema/http.graphqls:
```diff
@@ -1,4 +1,11 @@
+    """
+    http(
+        """
+        Http url to get the content from (e.g., "https://docs.dagger.io").
```
```suggestion
        HTTP URL to get the content from (e.g., "https://docs.dagger.io").
```

helderco reviewed (commented):
@TomChv, we're almost there, just a few quick cleanups 🚀

dolanor commented:
I haven't had time to review this, yet. I will do it tomorrow!

TomChv commented on core/schema/container.graphqls:
```diff
@@ -23,27 +25,35 @@ type Container {
     Image's address from its registry.
-    Formatted as {host}/{user}/{repo}:{tag} (e.g. docker.io/dagger/dagger:main).
+
+    Formatted as {host}/{user}/{repo}:{tag} (e.g., "docker.io/dagger/dagger:main").
```
I merged the PR, I'll take care of the rebase

TomChv commented:
@helderco comments applied and PR rebased with #4436 included!

There's only this comment: https://github.com/dagger/dagger/pull/4367#discussion_r1086619743 that I did not really understood.

dolanor commented on core/schema/git.graphqls:
```diff
@@ -1,24 +1,57 @@
+        """
+        Url of the git repository.
+        Can be formatted as https://{host}/{owner}/{repo}, git@{host}/{owner}/{repo}
+        Suffix ".git" is optional.
```
I don't think documenting git is not our problem. I would remove this line.

TomChv replied:
I think it's still useful for users to understand what kind of input they can provide.

dolanor replied:
Sorry, I was just talking about the last line (L9) about the `.git` suffix :)
Of course the string format expected is very useful and important here.

dolanor commented on core/schema/platform.graphqls:
```diff
@@ -1,12 +1,11 @@
 The platform config OS and architecture in a Container.
-The format is [os]/[platform]/[version] (e.g. darwin/arm64/v7, windows/amd64, linux/arm64).
+
+The format is [os]/[platform]/[version] (e.g. "darwin/arm64/v7", "windows/amd64", "linux/arm64").
```
```suggestion
The format is [os]/[platform]/[version] (e.g., "darwin/arm64/v7", "windows/amd64", "linux/arm64").
```

dolanor commented on core/schema/platform.graphqls:
```diff
@@ -1,12 +1,11 @@
-  """
-  The default platform of the builder.
-  """
+  "The default platform of the builder."
```
get back to `"""`

TomChv replied:
Why? It's a single line comment?

dolanor replied:
*https://github.com/dagger/dagger/pull/4367#issuecomment-1398350547*
> For one liners you can still use the triple quotes. Instead of doing it only when needed, I suggest making it a convention. It also helps with git diffs in case the content changes (e.g., turning it multiline or needing double quotes in an example).

So if we create some parser, it's always the same rule, we will parse a `"""` until the netxt `"""`. And for git diff, if we decide to add more description, the diff will just be in between those `"""`, and not switching from `"` to `"""`. So you can focus on what actually changed

```diff
  """
  Provides super interesting information
+ 
+ With this extra super information
+ about it.
  """
  ```
  
  Instead of 
```diff
- "Provides super interesting information"
+ """
+ Provides super interesting information
+ 
+ With this extra super information
+ about it.
+ """
  ```
  ```diff

dolanor replied:
But I won't block for that, I would do some PR dedicated to change every description into the `"""` so we can merge this one (which brings more goodness than bare `"""` ).

dolanor reviewed (changes requested):
I left a lot of comments that are mostly `"` -> `"""`. We should do an automated pass to clean up the all file for those. We could do that in a subsequential PR.

My only concern is removing the param names from the description. If we do that, it won't be idiomatic Go comment anymore, and since it's prose, we can't really add it after the fact.
And in the other hand, Go doesn't have a standard way to document those params.

TomChv commented:
> Also, did we create a PR/issue for the style guide?
 
This issue https://github.com/dagger/dagger/issues/4331 is related to a style enforced (linter) :D

TomChv commented on codegen/generator/nodejs/templates/functions.go:
```diff
@@ -60,6 +61,10 @@ func commentToLines(s string) []string {
 	return split
 }
 
+func getParamDescription(s []string) []string {
```
Make a difference between examples and doc, it's not useful for now but we may need it later so I kept it.

TomChv commented:
@dolanor I applied your comments in https://github.com/dagger/dagger/pull/4367/commits/954f7ffeab8d572c40c133203d6f02181e39e095, let me know if I have to update anything else.

dolanor commented:
I'm gonna fix the tests and merge as soon as I have Helder's feedback as well on the last points.

helderco commented on core/schema/container.graphqls:
```diff
@@ -61,121 +80,439 @@ type Container {
+    """
+    Permission given to the copied file (e.g., 0600).
+
+    Default: 0644.
```
Just a preference, not a blocker. I often see defaults in parenthesis after a longer description or as a [sentence after](https://docs.github.com/en/graphql/reference/queries#repository).

Example 1:

> Follow repository renames. If disabled, a repository referenced by its old name will return an error.
> **The default value is `true`.**

Example 2: 

> Follow repository renames. If disabled, a repository referenced by its old name will return an error.
> **Defaults to `true`.**

Example 3:

> Follow repository renames. If disabled, a repository referenced by its old name will return an error **(defaults to `true`)**.

By using a label it feels like there's special rendering for it, like in one language you could use `@default 0644`, in another it could be a markdown heading. Using prose is more neutral.

Again, it's just personal preference.

dolanor replied:
Ok, I got it. In GraphQL docs, it doesn't have a standard except a sentence.

I guess we could separate from the rest of the body of the description and if the line contains default, we extract the default value from the sentence. But we need to have a standard way to phrase it.
I guess `Default to <value>.` should be easily parseable and be standard enough.

TomChv commented:
> I'm gonna fix the tests and merge as soon as I have Helder's feedback as well on the last points.

PR rebased and tests fixed!

dolanor commented on codegen/generator/nodejs/templates/src/object_comment_test.go:
```diff
@@ -11,8 +11,7 @@ func TestComment(t *testing.T) {
 		tmpl := templateHelper(t)
-		want := `
-  /**
+		want := `/**
```
This is not good, it should be indented 1 time (2 spaces).

TomChv replied:
Fixed in https://github.com/dagger/dagger/pull/4367/commits/516b085d5e82abfa2961d86e42104adbf679134a

dolanor commented on codegen/generator/nodejs/templates/src/object_comment_test.go:
```diff
@@ -25,8 +24,7 @@ func TestComment(t *testing.T) {
 		tmpl := templateHelper(t)
-		want := `
-  /**
+		want := `/**
```
This is not good, it should be indented 1 time (2 spaces).

TomChv replied:
Fixed in https://github.com/dagger/dagger/pull/4367/commits/516b085d5e82abfa2961d86e42104adbf679134a

dolanor commented on codegen/generator/nodejs/templates/src/testdata/type_test_scalar_want.ts:
```diff
@@ -1,6 +1,5 @@
 
-
-  /**
+/**
```
This is not good, it should be indented 1 time (2 spaces).

TomChv replied:
Fixed in https://github.com/dagger/dagger/pull/4367/commits/516b085d5e82abfa2961d86e42104adbf679134a

dolanor reviewed (changes requested):
last tiny bits of indentation (both template code and generated code)

slumbering commented:
👋 Would it be possible to fix the latest Tanguy comments ? 
Once those are done I'll merge this PR @TomChv ;) Thanks
